KAFKA-4914: Partition re-assignment tool should check types before pe…

nicktrav opened the pull request:
…rsisting state in ZooKeeper

Prior to this, there have been instances where invalid data was allowed to be persisted in ZooKeeper, which causes ClassCastExceptions when a broker is restarted and reads this type-unsafe data.

Adds basic structural and type validation for the reassignment JSON via introduction of Scala case classes that map to the expected JSON structure.

nicktrav commented:
Prior to this, we have run into issues where invalid data was allowed to be persisted in ZooKeeper, which causes `ClassCastException`s when a broker is restarted and reads this type-unsafe data.

Adds basic structural and type validation for the reassignment JSON.

Also adds a check on `--execute` to ensure the reassignment JSON is still valid, given it can be edited after `--generate` and `--verify`.

nicktrav commented:
Hey @ijuma - would you mind taking a look? I see you've worked on this previously.

nicktrav commented:
@ewencp @ijuma - any chance of a review on this one?

omkreddy commented:
@nicktrav Thanks for the patch. LGTM. We also faced this issue in production. Can you rebase the PR?

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6169/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6153/
Test FAILed (JDK 8 and Scala 2.12).

nicktrav commented:
@omkreddy - thanks for checking this one out! I completely forgot that I had this open. I've rebased.

Would love to get it merged, as it was the source of a reasonable amount of (avoidable) pain for us a few months ago, and would love to save some others from that. Are you able to merge this in when green?

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6281/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6265/
Test PASSed (JDK 8 and Scala 2.12).

omkreddy commented:
LGTM.

Though we are deprecating "--zookeeper"  option in KIP-179 (Change ReassignPartitionsCommand to use adminClient), It is good to have this check.

cc @ijuma  pinging for review.

ijuma commented:
Thanks for the PR. I think we should consider using Jackson to parse into classes with the expected structure. That will do the necessary validation in a more robust way. Thoughts?

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6565/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6550/
Test FAILed (JDK 8 and Scala 2.12).

nicktrav commented:
Sounds reasonable, @ijuma. I'll take another pass at it.

Commit:
KAFKA-4914: Partition re-assignment tool should check types before persisting state in ZooKeeper

Prior to this, there have been instances where invalid data was allowed to be persisted in ZooKeeper, which causes ClassCastExceptions when a broker is restarted and reads this type-unsafe data.

Adds basic structural and type validation for the reassignment JSON via
introduction of Scala case classes that map to the expected JSON
structure.

nicktrav commented:
@ijuma - I've taken a different approach, as suggested, using case classes to hold the deserialized json. PTAL.

cc: @omkreddy

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6593/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6578/
Test PASSed (JDK 8 and Scala 2.12).

ijuma commented:
Thanks, I'll take a look tomorrow.

nicktrav commented:
@ijuma - sorry to keep pinging you on this. How is this looking? Happy to address any feedback you might have.

nicktrav commented:
Ping @omkreddy

omkreddy commented:
@nicktrav  Thanks for your patience.  committers are busy with 0.11.0.1 release and other tasks. I am sure this will get merged into 1.0.0/October release. Pls, rebase the PR when you get a chance.


cc @ijuma

nicktrav commented:
Retest this please.

nicktrav commented:
@omkreddy - rebased.

omkreddy commented:
cc @ijuma

ijuma commented on core/src/main/scala/kafka/utils/Json.scala:
```diff
@@ -59,4 +59,12 @@ object Json {
+  /**
+    * Parse a JSON string into either a generic type T, or a Throwable in the case of exception.
+    */
+  def parseTo[T](input: String, klass: Class[T]): Either[Throwable, T] = {
```
The idiomatic way of doing this is:

```scala
def parseTo[T](input: String)(implicit tag: ClassTag[T]): Either[Throwable, T] = {
```

Then the caller can just invoke `parseTo[PartitionAssignment](jsonData)`.

Also, should we call it `parseAs` perhaps?

ijuma commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -165,19 +165,19 @@ object ZkUtils {
+    if (parseResult.isLeft)
+      throw new ConfigException(s"Invalid reassignment config: ${parseResult.left}")
+
+    val assignments = parseResult.right
```
You should use pattern matching to retrieve `left` or `right` so that you can avoid the unsafe `get` method below.

ijuma commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -1177,3 +1177,25 @@ class ZKCheckedEphemeral(path: String,
+  *
+  * An assignment consists of a `version` and a list of `partitions`, which represent the assignment
+  * of topic-partitions to brokers.
+  */
```
I think we should not document these classes as being for deserialization only. We should also use them for serialization so that we detect problems at that stage too.

Also, we can move these classes to `ZkData`.

ijuma commented on core/src/test/scala/unit/kafka/utils/JsonTest.scala:
```diff
@@ -62,5 +65,26 @@ class JsonTest {
+
 }
+
+case class TestObject(@JsonProperty("foo") foo: String, @JsonProperty("bar") bar: Int)
```
It's better to place this inside a companion object so that it's restricted to this test.

ijuma commented on core/src/test/scala/unit/kafka/utils/JsonTest.scala:
```diff
@@ -62,5 +65,26 @@ class JsonTest {
+    val result = Json.parseTo(s"""{"foo": "$foo", "bar": $bar}""", classOf[TestObject])
+
+    assertTrue(result.isRight)
+    assertEquals(TestObject(foo, bar), result.right.get)
```
Maybe we can also have a test where we can roundtrip (serialize and then deserialize)? It would be nice to have a method in `Json` to serialize with `Jackson`. Maybe we can even actually replace the existing one.

ijuma commented on core/src/test/scala/unit/kafka/utils/ZkUtilsTest.scala:
```diff
@@ -84,4 +100,28 @@ class ZkUtilsTest extends ZooKeeperTestHarness {
   }
+
+  @Test
+  def testParsePartitionReassignmentDataWithoutDedup_invalidJson() = {
```
Nit: we don't use `_` in method names.

ijuma reviewed (commented):
Sorry for the long delay. Finally got around to reviewing this. Left a few comments.

Commit:
[feedback] parseAs

Commit:
[feedback] Pattern matching

Commit:
[feedback] Relocate POJOs and update documentation

Commit:
[feedback] JsonTest companion object

nicktrav commented:
Thanks for the review @ijuma! I've addressed your comments in separate commits.

For the case of your JSON parsing comment, I've added a new dependency on jackson-module-scala that deals with Scala collections a little better, rather than the existing custom parsing logic. Not sure how you feel about adding a new dependency, although there are already a couple of other Jackson libraries in use already.

ijuma commented:
Let's remove the last commit for now as there is an ongoing discussion on whether we should use jackson-scala (https://github.com/apache/kafka/pull/4259). The rest looks good.

nicktrav commented:
Done! Thanks for the review and getting this one over the line, @ijuma. 👍

ijuma committed:
Use PartitionAssignment/ReplicaAssignment in ReassignPartitionsZNode

ijuma commented:
@nicktrav, I had a closer look and it looks like this PR doesn't achieve the original goal as the Controller no longer uses `ZkUtils`. I think we should revert the `ZkUtils` changes and do something like https://github.com/nicktrav/kafka/pull/1

I didn't run the tests, but it gives you an idea of what I had in mind. Thoughts?

nicktrav commented:
@ijuma - sounds good! I didn't realize it had changed. Did you want to re-use the commit you pushed? Otherwise, I might add a few more unit tests to your changes and open a new PR.

ijuma commented:
You can just use whatever is useful in your PR. Thanks!

ijuma commented:
@nicktrav I just merged https://github.com/apache/kafka/commit/0a508a436c928c0440e8b90dda98c22dc0ec244c which converts to Jackson for serialization as well. Now would be a good time to submit your PR which would be the first one to transition from Maps to actual classes.

nicktrav commented:
@ijuma - awesome. Thanks for the follow-up. I'll try and get to this sometime this week.

Commit:
[rebase fixups]

nicktrav commented:
@ijuma - incorporated your changes and fixed up the remaining merge conflicts. This should be good for a final round of review. Thanks for persevering!

nicktrav commented:
Hey @ijuma - Happy New Year. Gentle ping on this one again. I've rebased this a few times now, so if you could take another look, I'd appreciate it. Thanks!

ijuma commented on core/src/main/scala/kafka/utils/Json.scala:
```diff
@@ -45,6 +46,14 @@ object Json {
+  /**
+   * Parse a JSON string into either a generic type T, or a JsonProcessingException in the case of exception.
+   */
+  def parseStringAs[T](input: String)(implicit tag: ClassTag[T]): Either[Throwable, T] = {
```
The result type can be `Either[JsonProcessingException, T]`.

ijuma commented on core/src/main/scala/kafka/zk/ZkData.scala:
```diff
@@ -534,3 +535,20 @@ object ZkData {
+  * represent the broker ids that the `TopicPartition` is assigned to.
+  */
+case class ReplicaAssignment(@JsonProperty("topic") topic: String,
+                             @JsonProperty("partition") partitions: Int,
```
Field name should be `partition`

ijuma commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -141,19 +142,20 @@ object ZkUtils {
+    val parseResult = Json.parseStringAs[PartitionAssignment](jsonData)
+
+    val assignments = parseResult match {
+      case Left(throwable) => throw new ConfigException(s"Invalid reassignment config: $throwable")
```
Seems like we'd want to include the original exception as the cause. Also, why is it a `ConfigException`?

ijuma commented on core/src/main/scala/kafka/zk/ZkData.scala:
```diff
@@ -534,3 +535,20 @@ object ZkData {
+  * An assignment consists of a `version` and a list of `partitions`, which represent the assignment
+  * of topic-partitions to brokers.
+  */
+case class PartitionAssignment(@JsonProperty("version") version: Int,
```
We should be using this in `ReassignPartitionsZNode` as in my PR.

ijuma commented on core/src/main/scala/kafka/utils/ZkUtils.scala:
```diff
@@ -141,19 +142,20 @@ object ZkUtils {
 
-  // Parses without deduplicating keys so the data can be checked before allowing reassignment to proceed
   def parsePartitionReassignmentData(jsonData: String): Map[TopicAndPartition, Seq[Int]] = {
+    val parseResult = Json.parseStringAs[PartitionAssignment](jsonData)
```
We should use `ReassignPartitionsZNode.decode`

ijuma commented on core/src/test/scala/unit/kafka/utils/ZkUtilsTest.scala:
```diff
@@ -107,4 +123,27 @@ class ZkUtilsTest extends ZooKeeperTestHarness {
+  }
+
+  @Test
+  def testParsePartitionReassignmentData() = {
```
We should rewrite these tests to use `ReassignPartitionsZNode`. Also, since they don't require `ZooKeeper`, they should be unit tests. Perhaps `ReassignPartitionsZNodeTest`.

ijuma reviewed (commented):
Sorry for the delay @nicktrav. I left a few comments.

Commit:
[feedback]

nicktrav commented:
Thanks for the review @ijuma. I pushed a new commit addressing your last round of comments. I'm sure some of it is not idiomatic Scala, so would appreciate some guidance, if any.

Commit:
[fixup] Encode bytes as UTF-8

Commit:
[fixup] Add missing license

omkreddy commented:
Can we use zkUtils.parsePartitionReassignmentData here
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala#L245
above change will solve KAFKA-6084. This can be done in another PR.

ijuma commented:
@omkreddy Ideally, we should not be using `ZkUtils` in `ReassignPartitionsCommand` (i.e. we should use `KafkaZkClient`). Also, `ReassignPartitionsCommand` also parses a `log_dirs` field which is not handled in `ZkUtils` (or `KafkaZkClient`). Let's do that clean-up in a separate PR.

nicktrav commented:
@omkreddy @ijuma - any action for me to take here based on the last two comments?

omkreddy commented:
LGTM. +1

ijuma commented:
Sorry, I have just moved countries and won't have time until next week.

nicktrav commented:
No worries at all. Good luck with the move!

nicktrav commented:
Any chance of an update on this @ijuma @omkreddy? (coincidentally, today is the 1 year birthday of the ticket! 🍰)

Commit:
[fixup] Revert unnecessary newline removal

omkreddy commented:
maybe @hachikuji  can help to merge this patch.

viktorsomogyi commented on core/src/main/scala/kafka/utils/Json.scala:
```diff
@@ -45,6 +46,15 @@ object Json {
+   * Parse a JSON string into either a generic type T, or a JsonProcessingException in the case of
+    * exception.
+   */
+  def parseStringAs[T](input: String)(implicit tag: ClassTag[T]): Either[JsonProcessingException, T] = {
```
Hmm, just a very basic question: why are you using here Either/Left/Right instead of Try/Success/Failure?
As from my previous experiences we used Try when we wanted to propagate the exception.
And please not I don't mind using Either, just asking for reasons. :)

nicktrav replied:
My (idiomatic) Scala isn't great, but this came at the suggestion of @ijuma (above):
https://github.com/apache/kafka/pull/2708#discussion_r151342320

ijuma replied:
`Try` doesn't let you specify an exception subclass and behaves a bit different if an exception is thrown during `map`, etc.

viktorsomogyi reviewed (approved):
Although (as mentioned on my JIRA) this PR seems to aim bit different goals then mine, I reviewed this PR and made one minor comment/question.
All in all I think the PR looks good. Would be happy to see this merged.

ijuma committed:
Minor fixes

ijuma reviewed (approved):
I pushed minor changes and it now LGTM. Will merge it after the tests pass. Thanks so much for your patience.

nicktrav commented:
Thanks @ijuma!
